sematic/api/endpoints/resolutions.py

```diff
@@ -345,5 +345,6 @@ def _cancel_non_terminal_runs(root_id):
         run.external_jobs = jobs
 
         run.future_state = FutureState.CANCELED
+        run.failed_at = datetime.utcnow()
```

Comment:
arguably, the run is not failed. Can we use `ended_at` instead?

Reply:
`ended_at` is when the execution of the func returned a `Future` instance.
This does not happen here, as there is not returned `Future`, and `ended_at` does not indicate reaching a terminal state; it will be followed up by `resolved_at`, or by `failed_at`.

Reply:
We can introduce a `canceled_at`, but it seems we have an explosion of timestamps.

Reply:
Ah my bad. Ok then `failed_at` is the most appropriate timestamp.

I agree, let's not add `canceled_at`. It is not scalable to add timestamps for each state. The longer-term solution is to keep a history of state changes (from_state, to_state, changed_at) either as a JSON field on `runs` or as a separate table, but that is work for another time.